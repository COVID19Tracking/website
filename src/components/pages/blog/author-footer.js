import React from 'react'
import AuthorFooterStyles from './author-footer.module.scss'

export default ({ authors }) => {
  const footerAuthors = authors.filter(
    author => author.childContentfulAuthorBiographyTextNode !== null,
  ) // only keep authors with biographies

  return (
    <div className={AuthorFooterStyles.container}>
      {footerAuthors.map(author => (
        <div
          className={AuthorFooterStyles.authorFooter}
          key={`author-${author.name}`}
        >
          {author.headshot && (
            <img
              src={author.headshot.resize.src}
              alt={author.headshot.file.fileName}
            />
          )}
          <div className={AuthorFooterStyles.textContent}>
            <div
              dangerouslySetInnerHTML={{
                __html:
                  author.childContentfulAuthorBiographyTextNode
                    .childMarkdownRemark.html,
              }}
            />
            <a href={`https://twitter.com/${author.twitterHandle}`}>
              @{author.twitterHandle}
            </a>
          </div>
        </div>
      ))}
    </div>
  )
}
