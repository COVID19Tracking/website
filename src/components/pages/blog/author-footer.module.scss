.container {
  margin: 2rem 0 0;
}

.author-footer {
  display: flex;
  flex-direction: row;
  img {
    width: 85px;
    height: 85px;
    flex-shrink: 0;
    margin-bottom: 1.5em; // matches layout content-primary p
  }
  .text-content {
    font-size: 16px;
    line-height: 1.75;
    font-style: italic;
    padding-left: .875rem;
    padding-right: 1.375rem;
    p {
      margin-bottom: 0;
    }
  }
}
