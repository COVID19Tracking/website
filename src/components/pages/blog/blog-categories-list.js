import React, { useState } from 'react'
import { useStaticQuery, graphql, Link } from 'gatsby'
import {
  TabletDisclosure,
  TabletDisclosureHeader,
  TabletDisclosureContent,
} from '~components/common/tablet-disclosure'
import blogCategoriesListStyles from './blog-categories-list.module.scss'

const BlogCategoriesList = () => {
  const { allContentfulBlogCategory } = useStaticQuery(graphql`
    query {
      allContentfulBlogCategory {
        nodes {
          name
          slug
          blog_post {
            id
          }
        }
      }
    }
  `)

  // put the categories with the most posts first
  const categories = allContentfulBlogCategory.nodes.sort((a, b) =>
    b.blog_post && a.blog_post ? b.blog_post.length - a.blog_post.length : 0,
  )

  const [isOpen, setIsOpen] = useState(false)

  return (
    <div className={blogCategoriesListStyles.wrapper}>
      <TabletDisclosure className={blogCategoriesListStyles.container}>
        <TabletDisclosureHeader isOpen={isOpen} setIsOpen={setIsOpen}>
          <h3>Posts by Category</h3>
        </TabletDisclosureHeader>
        <TabletDisclosureContent isOpen={isOpen}>
          <ul
            role="navigation"
            aria-label="Categories"
            className={blogCategoriesListStyles.categoryList}
          >
            {categories.map(category => (
              <li
                className={blogCategoriesListStyles.category}
                key={category.slug}
              >
                <Link to={`/blog/category/${category.slug}`}>
                  <p className={blogCategoriesListStyles.categoryText}>
                    {category.name} (
                    {category.blog_post && category.blog_post.length})
                  </p>
                </Link>
              </li>
            ))}
          </ul>
        </TabletDisclosureContent>
      </TabletDisclosure>
    </div>
  )
}

export default BlogCategoriesList
