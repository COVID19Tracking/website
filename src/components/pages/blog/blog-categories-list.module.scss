.wrapper {
  position: relative;
}

.container {
  border-radius: 4px;
  background-color: $color-plum-100;

  h3,
  a {
    @include type-size(100);
    color: $color-plum-700;
  }

  h3 {
    line-height: 1.6;
  }

  @media (min-width: $viewport-lg) {
    @include margin(24, left);
    min-width: 13rem;
  }
}

.category {
  @include padding(16, (left, right));
  @include padding(8, top);

  a {
    line-height: 0.8;

    .category-text {
      line-height: 1.4;
      @include margin(0);
    }
  }
}

.category {
  @include padding(8, top);
}

.category-list {
  border-top: 1px solid $color-plum-300;
  list-style-type: none;
  @include padding(0, left);
  @include margin(0, bottom);
  @include margin(8, top);

  li {
    @media (max-width: $viewport-lg) {
      font-size: 1rem;
      line-height: 2;
    }
    &:first-child {
      @include padding(8, top);
    }
  }

  &.opened {
    display: block;
  }

  @media (min-width: $viewport-lg) {
    display: block;
  }
}
