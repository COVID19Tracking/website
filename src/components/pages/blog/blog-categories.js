import React from 'react'
import { Link } from 'gatsby'
import classnames from 'classnames'
import categoryStyles from './blog-categories.module.scss'

const BlogCategories = ({ categories, lightBackground }) => {
  if (!categories) {
    return null
  }
  return (
    <div className={categoryStyles.categoriesContainer}>
      {categories.map(category => (
        <p
          className={classnames(
            categoryStyles.category,
            lightBackground && categoryStyles.lightBg,
          )}
          key={category.slug}
        >
          <Link to={`/blog/category/${category.slug}`}>{category.name}</Link>
        </p>
      ))}
    </div>
  )
}

export default BlogCategories
