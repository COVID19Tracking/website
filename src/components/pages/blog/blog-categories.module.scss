.categories-container {
  display: flex;

  @media (max-width: $viewport-sm) {
    flex-direction: column;
  }

  .category {
    font-weight: 600;
    @include type-size(100);
    line-height: 1rem;

    color: $color-plum-100;
    @include margin(0, bottom);

    @media (max-width: $viewport-sm) {
      @include margin(0, top);
    }

    a {
      text-decoration: none;
      &:hover {
        @include link-hover-dark-bg;
      }
      &:focus {
        @include link-focus-dark-bg;
      }
    }

    &::after {
      content: '  |  ';
      white-space: pre;
      color: $color-plum-100;
      font-weight: normal;

      @media (max-width: $viewport-sm) {
        content: '';
      }
    }

    &:last-child::after {
      content: none;
    }

    &.light-bg {
      color: $color-plum-700;
      a {
        text-decoration: none;
        &:hover {
          @include link-hover;
        }
        &:focus {
          @include link-focus;
        }
      }

      &::after {
        color: $color-plum-700;
      }
    }
  }
}
