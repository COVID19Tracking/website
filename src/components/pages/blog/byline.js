import React, { Fragment } from 'react'
import bylineStyles from './byline.module.scss'

const getAuthorLink = author => {
  if (author.link) {
    return author.link
  }
  if (author.twitterLink) {
    return author.twitterLink
  }
  return null
}

const getBylineClass = (isSmallMargin, hasImage) => {
  if (isSmallMargin) {
    if (hasImage) {
      return bylineStyles.bylineSmallMarginImage
    }
    return bylineStyles.bylineSmallMargin
  }
  return bylineStyles.byline
}

const Author = ({ author }) => {
  const link = getAuthorLink(author)
  return (
    <>
      {link ? (
        <a className={bylineStyles.authorInner} href={link}>
          {author.name}
        </a>
      ) : (
        <span className={bylineStyles.authorInner}>{author.name}</span>
      )}
    </>
  )
}

const AuthorSpacer = ({ index, length }) => {
  if (index === 0 && length === 2) {
    return <> & </> // first author, only two items in the list
  }
  if (index === length - 2) {
    return <>, & </> // second to last author
  }
  if (index === length - 1) {
    return null // last author
  }
  return <>, </>
}

const AuthorsText = ({ authors }) => (
  <>
    {authors.map((author, index) => (
      <Fragment key={`author-${author.name}`}>
        <Author author={author} />
        <AuthorSpacer index={index} length={authors.length} />
      </Fragment>
    ))}
  </>
)

const Byline = ({ authors, date, smallmargin = false }) => {
  const authorsWithHeadshots = authors.filter(author => author.headshot)
  const hasHeadshots = authorsWithHeadshots.length > 0
  const bylineClass = getBylineClass(smallmargin, hasHeadshots)
  return (
    <div className={bylineClass}>
      {hasHeadshots && (
        <div className={bylineStyles.headshotContainer}>
          {authorsWithHeadshots.map(author => (
            <img
              src={author.headshot.resize.src}
              alt={author.headshot.file.fileName}
            />
          ))}
        </div>
      )}
      <p className={bylineStyles.authorDateContainer}>
        By{' '}
        <p className={bylineStyles.author}>
          <AuthorsText authors={authors} />
        </p>
        <span className={bylineStyles.date}>{date}</span>
      </p>
    </div>
  )
}

const OneLineByline = ({ authors, date }) => (
  <p className={bylineStyles.oneLineByline}>
    By{' '}
    <span className={bylineStyles.author}>
      <AuthorsText authors={authors} />
    </span>
    <span className={bylineStyles.relatedDate}>{date}</span>
  </p>
)

export { Byline, AuthorsText, OneLineByline }
