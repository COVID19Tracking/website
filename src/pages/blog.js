import React from 'react'
import { graphql } from 'gatsby'
import Layout from '../components/layout'
import BlogTeaserList from '../components/pages/blog/blog-teaser-list'

export default ({ data }) => (
  <Layout title="Blog" textHeavy narrow>
    <BlogTeaserList items={data.allContentfulBlogPost.edges} />
  </Layout>
)

export const query = graphql`
  query {
    allContentfulBlogPost(sort: { fields: publishDate, order: DESC }) {
      edges {
        node {
          title
          slug
          authors {
            name
            twitterLink
            link
            headshot {
              file {
                fileName
              }
              resize(width: 100) {
                width
                height
                src
              }
            }
          }
          publishDate(formatString: "MMMM D, YYYY")
          lede {
            lede
          }
        }
      }
    }
  }
`
