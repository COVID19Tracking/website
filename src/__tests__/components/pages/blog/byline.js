import React from 'react'
import renderer from 'react-test-renderer'
import { Byline } from '../../../../components/pages/blog/byline'

describe('Components : Pages : Blog : Byline', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Byline
          authors={[
            {
              name: 'test name',
              twitterLink: 'https://twitter.com/test',
              headshot: {
                file: {
                  fileName: 'file.png',
                },
                resize: {
                  src: '//images.ctfassets.net/arandomfile',
                },
              },
            },
          ]}
          date="Aug 1 2020"
        />,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()

    const treeNoLink = renderer
      .create(
        <Byline
          authors={[
            {
              name: 'test name',
            },
          ]}
          date="Aug 1 2020"
        />,
      )
      .toJSON()
    expect(treeNoLink).toMatchSnapshot()
  })
})
