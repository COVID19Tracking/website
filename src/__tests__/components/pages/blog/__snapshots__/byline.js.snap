// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components : Pages : Blog : Byline renders correctly 1`] = `
<div
  className="byline"
>
  <div
    className="headshotContainer"
  >
    <img
      alt="file.png"
      src="//images.ctfassets.net/arandomfile"
    />
  </div>
  <p
    className="authorDateContainer"
  >
    By
     
    <p
      className="author"
    >
      <a
        className="authorInner"
        href="https://twitter.com/test"
      >
        test name
      </a>
    </p>
    <span
      className="date"
    >
      Aug 1 2020
    </span>
  </p>
</div>
`;

exports[`Components : Pages : Blog : Byline renders correctly 2`] = `
<div
  className="byline"
>
  <p
    className="authorDateContainer"
  >
    By
     
    <p
      className="author"
    >
      <span
        className="authorInner"
      >
        test name
      </span>
    </p>
    <span
      className="date"
    >
      Aug 1 2020
    </span>
  </p>
</div>
`;
